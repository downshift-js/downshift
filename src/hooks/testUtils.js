import {act} from '@testing-library/react'

export const items = [
  'Neptunium',
  'Plutonium',
  'Americium',
  'Curium',
  'Berkelium',
  'Californium',
  'Einsteinium',
  'Fermium',
  'Mendelevium',
  'Nobelium',
  'Lawrencium',
  'Rutherfordium',
  'Dubnium',
  'Seaborgium',
  'Bohrium',
  'Hassium',
  'Meitnerium',
  'Darmstadtium',
  'Roentgenium',
  'Copernicium',
  'Nihonium',
  'Flerovium',
  'Moscovium',
  'Livermorium',
  'Tennessine',
  'Oganesson',
]

export const defaultIds = {
  labelId: 'downshift-test-id-label',
  menuId: 'downshift-test-id-menu',
  getItemId: index => `downshift-test-id-item-${index}`,
  toggleButtonId: 'downshift-test-id-toggle-button',
  inputId: 'downshift-test-id-input',
}

export const waitForDebouncedA11yStatusUpdate = () => {
  act(() => jest.advanceTimersByTime(200))
}
