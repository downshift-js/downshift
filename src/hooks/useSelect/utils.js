import PropTypes from 'prop-types'
import {getNextWrappingIndex} from '../utils'

const defaultStateValues = {
  highlightedIndex: -1,
  isOpen: false,
  selectedItem: null,
}

function getA11yStatusMessage({isOpen, items}) {
  if (!items) {
    return ''
  }
  const resultCount = items.length
  if (isOpen) {
    if (resultCount === 0) {
      return 'No results are available'
    }
    return `${resultCount} result${
      resultCount === 1 ? ' is' : 's are'
    } available, use up and down arrow keys to navigate. Press Enter key to select.`
  }
  return ''
}

function getA11ySelectionMessage({selectedItem, itemToString}) {
  return `${itemToString(selectedItem)} has been selected.`
}

function getHighlightedIndexOnOpen(props, state, offset) {
  const {items, initialHighlightedIndex, defaultHighlightedIndex} = props
  const {selectedItem, highlightedIndex} = state

  // initialHighlightedIndex will give value to highlightedIndex on initial state only.
  if (initialHighlightedIndex !== undefined && highlightedIndex > -1) {
    return initialHighlightedIndex
  }
  if (defaultHighlightedIndex !== undefined) {
    return defaultHighlightedIndex
  }
  if (selectedItem) {
    if (offset === 0) {
      return items.indexOf(selectedItem)
    }
    return getNextWrappingIndex(
      offset,
      items.indexOf(selectedItem),
      items.length,
      false,
    )
  }
  if (offset === 0) {
    return -1
  }
  return offset < 0 ? items.length - 1 : 0
}

function capitalizeString(string) {
  return `${string.slice(0, 1).toUpperCase()}${string.slice(1)}`
}

function getDefaultValue(props, propKey) {
  const defaultPropKey = `default${capitalizeString(propKey)}`
  if (props[defaultPropKey] !== undefined) {
    return props[defaultPropKey]
  }
  return defaultStateValues[propKey]
}

function getInitialValue(props, propKey) {
  if (props[propKey] !== undefined) {
    return props[propKey]
  }
  const initialPropKey = `initial${capitalizeString(propKey)}`
  if (props[initialPropKey] !== undefined) {
    return props[initialPropKey]
  }
  return getDefaultValue(props, propKey)
}

function getInitialState(props) {
  return {
    highlightedIndex: getInitialValue(props, 'highlightedIndex'),
    isOpen: getInitialValue(props, 'isOpen'),
    selectedItem: getInitialValue(props, 'selectedItem'),
    keysSoFar: '',
  }
}

function invokeOnChangeHandler(propKey, props, state, changes) {
  const handler = `on${capitalizeString(propKey)}Change`
  if (
    props[handler] &&
    changes[propKey] !== undefined &&
    changes[propKey] !== state[propKey]
  ) {
    props[handler](changes)
  }
}

function callOnChangeProps(props, state, changes) {
  ;['isOpen', 'highlightedIndex', 'selectedItem'].forEach(propKey => {
    invokeOnChangeHandler(propKey, props, state, changes)
  })

  if (props.onStateChange && changes !== undefined) {
    props.onStateChange(changes)
  }
}

const propTypes = {
  items: PropTypes.array.isRequired,
  itemToString: PropTypes.func,
  getA11yStatusMessage: PropTypes.func,
  getA11ySelectionMessage: PropTypes.func,
  circularNavigation: PropTypes.bool,
  highlightedIndex: PropTypes.number,
  defaultHighlightedIndex: PropTypes.number,
  initialHighlightedIndex: PropTypes.number,
  isOpen: PropTypes.bool,
  defaultIsOpen: PropTypes.bool,
  initialIsOpen: PropTypes.bool,
  selectedItem: PropTypes.any,
  initialSelectedItem: PropTypes.any,
  defaultSelectedItem: PropTypes.any,
  id: PropTypes.string,
  labelId: PropTypes.string,
  menuId: PropTypes.string,
  getItemId: PropTypes.func,
  toggleButtonId: PropTypes.string,
  stateReducer: PropTypes.func,
  onSelectedItemChange: PropTypes.func,
  onHighlightedIndexChange: PropTypes.func,
  onStateChange: PropTypes.func,
  onIsOpenChange: PropTypes.func,
  environment: PropTypes.shape({
    addEventListener: PropTypes.func,
    removeEventListener: PropTypes.func,
    document: PropTypes.shape({
      getElementById: PropTypes.func,
      activeElement: PropTypes.any,
      body: PropTypes.any,
    }),
  }),
}

export {
  getHighlightedIndexOnOpen,
  getA11yStatusMessage,
  getA11ySelectionMessage,
  getInitialState,
  defaultStateValues,
  propTypes,
  getDefaultValue,
  callOnChangeProps,
}
