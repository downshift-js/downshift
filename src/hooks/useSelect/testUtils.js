import React from 'react'
import {useId} from '@reach/auto-id'
import {render} from '@testing-library/react'
import {renderHook} from '@testing-library/react-hooks'
import {getElementIds, itemToString} from '../utils'
import useSelect from '.'

const items = [
  'Neptunium',
  'Plutonium',
  'Americium',
  'Curium',
  'Berkelium',
  'Californium',
  'Einsteinium',
  'Fermium',
  'Mendelevium',
  'Nobelium',
  'Lawrencium',
  'Rutherfordium',
  'Dubnium',
  'Seaborgium',
  'Bohrium',
  'Hassium',
  'Meitnerium',
  'Darmstadtium',
  'Roentgenium',
  'Copernicium',
  'Nihonium',
  'Flerovium',
  'Moscovium',
  'Livermorium',
  'Tennessine',
  'Oganesson',
]

jest.mock('@reach/auto-id', () => {
  return {
    useId: () => 'test-id',
  }
})

const defaultIds = getElementIds(useId)

const dataTestIds = {
  toggleButton: 'toggle-button-id',
  menu: 'menu-id',
  item: index => `item-id-${index}`,
}

const setupHook = props => {
  return renderHook(() => useSelect({items, ...props}))
}

const DropdownSelect = props => {
  const {
    isOpen,
    selectedItem,
    getToggleButtonProps,
    getLabelProps,
    getMenuProps,
    highlightedIndex,
    getItemProps,
  } = useSelect({items, ...props})
  return (
    <div>
      <label {...getLabelProps()}>Choose an element:</label>
      <button
        data-testid={dataTestIds.toggleButton}
        {...getToggleButtonProps()}
      >
        {(selectedItem && selectedItem instanceof Object
          ? itemToString(selectedItem)
          : selectedItem) || 'Elements'}
      </button>
      <ul data-testid={dataTestIds.menu} {...getMenuProps()}>
        {isOpen &&
          (props.items || items).map((item, index) => {
            const stringItem =
              item instanceof Object ? itemToString(item) : item
            return (
              <li
                data-testid={dataTestIds.item(index)}
                style={
                  highlightedIndex === index ? {backgroundColor: 'blue'} : {}
                }
                key={`${stringItem}${index}`}
                {...getItemProps({item, index})}
              >
                {stringItem}
              </li>
            )
          })}
      </ul>
    </div>
  )
}

const setup = props => render(<DropdownSelect {...props} />)

export {dataTestIds, setup, items, setupHook, defaultIds}
