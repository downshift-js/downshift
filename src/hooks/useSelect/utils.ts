import PropTypes from 'prop-types'
import {defaultProps as commonDefaultProps} from '../utils'
import {noop} from '../../utils'
import {A11yStatusMessageOptions} from '../../types'
import {GetItemIndexByCharacterKeyOptions} from './types'

export function getItemIndexByCharacterKey<Item>({
  keysSoFar,
  highlightedIndex,
  items,
  itemToString,
  getItemNodeFromIndex,
}: GetItemIndexByCharacterKeyOptions<Item>) {
  const lowerCasedKeysSoFar = keysSoFar.toLowerCase()

  for (let index = 0; index < items.length; index++) {
    const offsetIndex = (index + highlightedIndex + 1) % items.length

    if (
      itemToString(items[offsetIndex] ?? null)
        .toLowerCase()
        .startsWith(lowerCasedKeysSoFar)
    ) {
      const element = getItemNodeFromIndex(offsetIndex)

      if (!element?.hasAttribute('disabled')) {
        return offsetIndex
      }
    }
  }

  return highlightedIndex
}

const propTypes = {
  items: PropTypes.array.isRequired,
  itemToString: PropTypes.func,
  getA11yStatusMessage: PropTypes.func,
  getA11ySelectionMessage: PropTypes.func,
  circularNavigation: PropTypes.bool,
  highlightedIndex: PropTypes.number,
  defaultHighlightedIndex: PropTypes.number,
  initialHighlightedIndex: PropTypes.number,
  isOpen: PropTypes.bool,
  defaultIsOpen: PropTypes.bool,
  initialIsOpen: PropTypes.bool,
  selectedItem: PropTypes.any,
  initialSelectedItem: PropTypes.any,
  defaultSelectedItem: PropTypes.any,
  id: PropTypes.string,
  labelId: PropTypes.string,
  menuId: PropTypes.string,
  getItemId: PropTypes.func,
  toggleButtonId: PropTypes.string,
  stateReducer: PropTypes.func,
  onSelectedItemChange: PropTypes.func,
  onHighlightedIndexChange: PropTypes.func,
  onStateChange: PropTypes.func,
  onIsOpenChange: PropTypes.func,
  environment: PropTypes.shape({
    addEventListener: PropTypes.func,
    removeEventListener: PropTypes.func,
    document: PropTypes.shape({
      getElementById: PropTypes.func,
      activeElement: PropTypes.any,
      body: PropTypes.any,
    }),
  }),
}

/**
 * Default implementation for status message. Only added when menu is open.
 * Will specift if there are results in the list, and if so, how many,
 * and what keys are relevant.
 *
 * @param {Object} param the downshift state and other relevant properties
 * @return {String} the a11y status message
 */
function getA11yStatusMessage<Item>({
  isOpen,
  resultCount,
  previousResultCount,
}: A11yStatusMessageOptions<Item>): string {
  if (!isOpen) {
    return ''
  }

  if (!resultCount) {
    return 'No results are available.'
  }

  if (resultCount !== previousResultCount) {
    return `${resultCount} result${
      resultCount === 1 ? ' is' : 's are'
    } available, use up and down arrow keys to navigate. Press Enter or Space Bar keys to select.`
  }

  return ''
}

export const defaultProps = {
  ...commonDefaultProps,
  getA11yStatusMessage,
}

// eslint-disable-next-line import/no-mutable-exports
export let validatePropTypes = noop as (
  options: unknown,
  caller: Function,
) => void
/* istanbul ignore next */
if (process.env.NODE_ENV !== 'production') {
  validatePropTypes = (options: unknown, caller: Function): void => {
    PropTypes.checkPropTypes(propTypes, options, 'prop', caller.name)
  }
}
