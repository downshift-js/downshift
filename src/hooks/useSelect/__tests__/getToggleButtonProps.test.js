/* eslint-disable jest/no-disabled-tests */
import keyboardKey from 'keyboard-key'
import {fireEvent, cleanup} from '@testing-library/react'
import {act as rtlAct} from '@testing-library/react-hooks'
import {act} from 'react-dom/test-utils'
import {noop} from '../../../utils'
import {setup, dataTestIds, items, setupHook, defaultIds} from '../testUtils'

describe('getToggleButtonProps', () => {
  afterEach(cleanup)

  describe('hook props', () => {
    test('assign default value to aria-labelledby', () => {
      const {result} = setupHook()
      const toggleButtonProps = result.current.getToggleButtonProps()

      expect(toggleButtonProps['aria-labelledby']).toEqual(
        `${defaultIds.labelId} ${defaultIds.toggleButtonId}`,
      )
    })

    test('assign custom value passed by user to aria-labelledby', () => {
      const props = {
        labelId: 'my-custom-label-id',
        toggleButtonId: 'my-custom-toggle-button-id',
      }
      const {result} = setupHook(props)
      const toggleButtonProps = result.current.getToggleButtonProps()

      expect(toggleButtonProps['aria-labelledby']).toEqual(
        `${props.labelId} ${props.toggleButtonId}`,
      )
    })

    test('assign default value to id', () => {
      const {result} = setupHook()
      const toggleButtonProps = result.current.getToggleButtonProps()

      expect(toggleButtonProps.id).toEqual(defaultIds.toggleButtonId)
    })

    test('assign custom value passed by user to id', () => {
      const props = {
        toggleButtonId: 'my-custom-toggle-button-id',
      }
      const {result} = setupHook(props)
      const toggleButtonProps = result.current.getToggleButtonProps()

      expect(toggleButtonProps.id).toEqual(props.toggleButtonId)
    })

    test("assign 'listbbox' to aria-haspopup", () => {
      const {result} = setupHook()
      const toggleButtonProps = result.current.getToggleButtonProps()

      expect(toggleButtonProps['aria-haspopup']).toEqual('listbox')
    })

    test("assign 'false' value to aria-expanded when menu is closed", () => {
      const {result} = setupHook({isOpen: false})
      const toggleButtonProps = result.current.getToggleButtonProps()

      expect(toggleButtonProps['aria-expanded']).toEqual(false)
    })

    test("assign 'true' value to aria-expanded when menu is open", () => {
      const {result} = setupHook()

      rtlAct(() => {
        const {ref: menuRef} = result.current.getMenuProps()

        menuRef({focus: noop})
        result.current.toggleMenu()
      })

      const toggleButtonProps = result.current.getToggleButtonProps()

      expect(toggleButtonProps['aria-expanded']).toEqual(true)
    })
  })

  describe('user props', () => {
    test('are passed down', () => {
      const {result} = setupHook()

      expect(result.current.getToggleButtonProps({foo: 'bar'})).toHaveProperty(
        'foo',
        'bar',
      )
    })

    test('event handler onClick is called along with downshift handler', () => {
      const userOnClick = jest.fn()
      const {result} = setupHook()

      rtlAct(() => {
        const {ref: menuRef} = result.current.getMenuProps()
        const {
          ref: toggleButtonRef,
          onClick,
        } = result.current.getToggleButtonProps({onClick: userOnClick})

        menuRef({focus: noop})
        toggleButtonRef({})
        onClick({})
      })

      expect(userOnClick).toHaveBeenCalledTimes(1)
      expect(result.current.isOpen).toBe(true)
    })

    test('event handler onKeyDown is called along with downshift handler', () => {
      const userOnKeyDown = jest.fn()
      const {result} = setupHook()

      rtlAct(() => {
        const {ref: menuRef} = result.current.getMenuProps()
        const {
          ref: toggleButtonRef,
          onKeyDown,
        } = result.current.getToggleButtonProps({onKeyDown: userOnKeyDown})

        menuRef({focus: noop})
        toggleButtonRef({})
        onKeyDown({keyCode: keyboardKey.ArrowDown, preventDefault: noop})
      })

      expect(userOnKeyDown).toHaveBeenCalledTimes(1)
      expect(result.current.isOpen).toBe(true)
    })

    test("event handler onClick is called without downshift handler if 'preventDownshiftDefault' is passed in user event", () => {
      const userOnClick = jest.fn(event => {
        event.preventDownshiftDefault = true
      })
      const {result} = setupHook()

      rtlAct(() => {
        const {ref: menuRef} = result.current.getMenuProps()
        const {
          ref: toggleButtonRef,
          onClick,
        } = result.current.getToggleButtonProps({onClick: userOnClick})

        toggleButtonRef({focus: noop})
        menuRef({focus: noop})
        onClick({})
      })

      expect(userOnClick).toHaveBeenCalledTimes(1)
      expect(result.current.isOpen).toBe(false)
    })

    test("event handler onKeyDown is called without downshift handler if 'preventDownshiftDefault' is passed in user event", () => {
      const userOnKeyDown = jest.fn(event => {
        event.preventDownshiftDefault = true
      })
      const {result} = setupHook()

      rtlAct(() => {
        const {ref: menuRef} = result.current.getMenuProps()
        const {
          ref: toggleButtonRef,
          onKeyDown,
        } = result.current.getToggleButtonProps({onKeyDown: userOnKeyDown})

        toggleButtonRef({focus: noop})
        menuRef({focus: noop})
        onKeyDown({
          keyCode: keyboardKey.ArrowDown,
          preventDefault: noop,
        })
      })

      expect(userOnKeyDown).toHaveBeenCalledTimes(1)
      expect(result.current.isOpen).toBe(false)
    })
  })

  describe('event handlers', () => {
    describe('on click', () => {
      test('opens the closed menu', () => {
        const wrapper = setup()
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(menu.childNodes).toHaveLength(items.length)
      })

      test('closes the open menu', () => {
        const wrapper = setup({initialIsOpen: true})
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(menu.childNodes).toHaveLength(0)
      })

      test('opens the closed menu without any option highlighted', () => {
        const wrapper = setup()
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBeNull()
      })

      test('opens the closed menu with selected option highlighted', () => {
        const selectedIndex = 3
        const wrapper = setup({initialSelectedItem: items[selectedIndex]})
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBe(
          defaultIds.getItemId(selectedIndex),
        )
      })

      test('opens the closed menu at initialHighlightedIndex, but on first click only', () => {
        const initialHighlightedIndex = 3
        const wrapper = setup({initialHighlightedIndex})
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBe(
          defaultIds.getItemId(initialHighlightedIndex),
        )

        fireEvent.click(toggleButton)
        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBeNull()
      })

      test('opens the closed menu at defaultHighlightedIndex, on every click', () => {
        const defaultHighlightedIndex = 3
        const wrapper = setup({defaultHighlightedIndex})
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBe(
          defaultIds.getItemId(defaultHighlightedIndex),
        )

        fireEvent.click(toggleButton)
        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBe(
          defaultIds.getItemId(defaultHighlightedIndex),
        )
      })

      test('opens the closed menu at highlightedIndex from props, on every click', () => {
        const highlightedIndex = 3
        const wrapper = setup({highlightedIndex})
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBe(
          defaultIds.getItemId(highlightedIndex),
        )

        fireEvent.click(toggleButton)
        fireEvent.click(toggleButton)

        expect(menu.getAttribute('aria-activedescendant')).toBe(
          defaultIds.getItemId(highlightedIndex),
        )
      })

      test('opens the closed menu and sets focus on the menu', () => {
        const wrapper = setup()
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.click(toggleButton)

        expect(document.activeElement).toBe(menu)
      })

      test('closes the open menu and sets focus on the toggle button', () => {
        const wrapper = setup({initialIsOpen: true})
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

        fireEvent.click(toggleButton)

        expect(document.activeElement).toBe(toggleButton)
      })
    })

    describe('on keydown', () => {
      describe('character key', () => {
        jest.useFakeTimers()

        afterEach(() => {
          act(() => jest.runAllTimers())
        })

        const startsWithCharacter = (option, character) => {
          return option.toLowerCase().startsWith(character.toLowerCase())
        }

        test('should select the first item that starts with that key', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {key: 'c'})

          expect(toggleButton.textContent).toEqual(
            items[items.findIndex(option => startsWithCharacter(option, 'c'))],
          )
        })

        test('should select the second item that starts with that key after typing it twice', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const firstIndex = items.findIndex(option =>
            startsWithCharacter(option, 'c'),
          )

          fireEvent.keyDown(toggleButton, {key: 'c'})
          act(() => jest.runAllTimers())
          fireEvent.keyDown(toggleButton, {key: 'c'})

          expect(toggleButton.textContent).toEqual(
            items[
              firstIndex +
                1 +
                items
                  .slice(firstIndex + 1)
                  .findIndex(option => startsWithCharacter(option, 'c'))
            ],
          )
        })

        test('should select the first item again if the items are depleated', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {key: 'b'})
          act(() => jest.runAllTimers())
          fireEvent.keyDown(toggleButton, {key: 'b'})
          act(() => jest.runAllTimers())
          fireEvent.keyDown(toggleButton, {key: 'b'})

          expect(toggleButton.textContent).toEqual(
            items[items.findIndex(option => startsWithCharacter(option, 'b'))],
          )
        })

        test('should not select anything if no item starts with that key', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {key: 'x'})

          expect(toggleButton.textContent).toEqual('Elements')
        })

        test('should select the first item that starts with the keys typed in rapid succession', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {key: 'c'})
          fireEvent.keyDown(toggleButton, {key: 'a'})

          expect(toggleButton.textContent).toEqual(
            items[items.findIndex(option => startsWithCharacter(option, 'ca'))],
          )
        })

        test('should become first character after timeout passes', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {key: 'c'})
          fireEvent.keyDown(toggleButton, {key: 'a'})
          act(() => jest.runAllTimers())
          fireEvent.keyDown(toggleButton, {key: 'l'})

          expect(toggleButton.textContent).toEqual(
            items[items.findIndex(option => startsWithCharacter(option, 'l'))],
          )
        })

        /* Here we just want to make sure the keys cleanup works. */
        test('should not go to the second option starting with the key if timeout did not pass', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {key: 'l'})
          act(() => jest.advanceTimersByTime(200)) // wait some time but not enough to trigger debounce.
          fireEvent.keyDown(toggleButton, {key: 'l'})
          act(() => jest.advanceTimersByTime(200)) // wait some time but not enough to trigger debounce.
          fireEvent.keyDown(toggleButton, {key: 'l'})
          act(() => jest.advanceTimersByTime(200)) // wait some time but not enough to trigger debounce.
          fireEvent.keyDown(toggleButton, {key: 'l'})

          // highlight should stay on the first item starting with 'L'
          expect(toggleButton.textContent).toEqual(
            items[items.findIndex(option => startsWithCharacter(option, 'l'))],
          )
        })
      })

      describe('arrow up', () => {
        test('opens the closed menu with last option highlighted', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowUp})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(items.length - 1),
          )
        })

        test('opens the closed menu with selected option - 1 highlighted', () => {
          const selectedIndex = 3
          const wrapper = setup({initialSelectedItem: items[selectedIndex]})
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowUp})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(selectedIndex - 1),
          )
        })

        test('opens the closed menu at initialHighlightedIndex, but on first arrow up only', () => {
          const initialHighlightedIndex = 3
          const wrapper = setup({initialHighlightedIndex})
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowUp})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(initialHighlightedIndex),
          )

          fireEvent.keyDown(menu, {keyCode: keyboardKey.Escape})
          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowUp})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(items.length - 1),
          )
        })

        test('arrow up opens the closed menu at defaultHighlightedIndex, on every arrow up', () => {
          const defaultHighlightedIndex = 3
          const wrapper = setup({defaultHighlightedIndex})
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowUp})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(defaultHighlightedIndex),
          )

          fireEvent.keyDown(menu, {keyCode: keyboardKey.Escape})
          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowUp})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(defaultHighlightedIndex),
          )
        })

        test.skip('prevents event default', () => {
          const wrapper = setup()
          const preventDefault = jest.fn()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {
            keyCode: keyboardKey.ArrowUp,
            preventDefault,
          })

          expect(preventDefault).toHaveBeenCalledTimes(1)
        })

        test('opens the closed menu and focuses the list', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowUp})

          expect(document.activeElement).toBe(menu)
        })
      })

      describe('arrow down', () => {
        test('opens the closed menu with first option highlighted', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowDown})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(0),
          )
        })

        test('opens the closed menu with selected option + 1 highlighted', () => {
          const selectedIndex = 3
          const wrapper = setup({initialSelectedItem: items[selectedIndex]})
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowDown})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(selectedIndex + 1),
          )
        })

        test('opens the closed menu at initialHighlightedIndex, but on first arrow down only', () => {
          const initialHighlightedIndex = 3
          const wrapper = setup({initialHighlightedIndex})
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowDown})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(initialHighlightedIndex),
          )

          fireEvent.keyDown(menu, {keyCode: keyboardKey.Escape})
          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowDown})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(0),
          )
        })

        test('opens the closed menu at defaultHighlightedIndex, on every arrow down', () => {
          const defaultHighlightedIndex = 3
          const wrapper = setup({defaultHighlightedIndex})
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowDown})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(defaultHighlightedIndex),
          )

          fireEvent.keyDown(menu, {keyCode: keyboardKey.Escape})
          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowDown})

          expect(menu.getAttribute('aria-activedescendant')).toBe(
            defaultIds.getItemId(defaultHighlightedIndex),
          )
        })

        // also add for menu when this test works.
        test.skip('arrow down prevents event default', () => {
          const wrapper = setup()
          const preventDefault = jest.fn()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)

          fireEvent.keyDown(toggleButton, {
            keyCode: keyboardKey.ArrowDown,
            preventDefault,
          })

          expect(preventDefault).toHaveBeenCalledTimes(1)
        })

        test('opens the closed menu and focuses the list', () => {
          const wrapper = setup()
          const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
          const menu = wrapper.getByTestId(dataTestIds.menu)

          fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.ArrowDown})

          expect(document.activeElement).toBe(menu)
        })
      })

      test("other than te ones supported don't affect anything", () => {
        const wrapper = setup()
        const toggleButton = wrapper.getByTestId(dataTestIds.toggleButton)
        const menu = wrapper.getByTestId(dataTestIds.menu)

        fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.Alt})
        fireEvent.keyDown(toggleButton, {keyCode: keyboardKey.Control})

        expect(toggleButton.textContent).toEqual('Elements')
        expect(menu.getAttribute('aria-activedescendant')).toBeNull()
        expect(menu.childNodes).toHaveLength(0)
      })
    })
  })
})
