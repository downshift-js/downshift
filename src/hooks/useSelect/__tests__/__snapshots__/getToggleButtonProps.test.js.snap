// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getToggleButtonProps event handlers on keydown arrow down defaultHighlightedIndex is ignored if item is disabled 1`] = `
[
  [
    Americium,
    2,
  ],
  [
    Americium,
    2,
  ],
  [
    Neptunium,
    0,
  ],
]
`;

exports[`getToggleButtonProps event handlers on keydown arrow down initialHighlightedIndex is ignored if item is disabled 1`] = `
[
  [
    Americium,
    2,
  ],
  [
    Neptunium,
    0,
  ],
]
`;

exports[`getToggleButtonProps event handlers on keydown arrow up defaultHighlightedIndex is ignored if item is disabled 1`] = `
[
  [
    Americium,
    2,
  ],
  [
    Americium,
    2,
  ],
  [
    Oganesson,
    25,
  ],
]
`;

exports[`getToggleButtonProps event handlers on keydown arrow up initialHighlightedIndex is ignored if item is disabled 1`] = `
[
  [
    Americium,
    2,
  ],
  [
    Oganesson,
    25,
  ],
]
`;
